model/metrics/messaging-metrics.yaml

```
@@ -0,0 +1,80 @@
+    brief: "Common messaging metrics attributes."
+    extends: attributes.messaging.common.minimal
+    attributes:
+      - ref: messaging.system
+        requirement_level: required
+      - ref: messaging.destination.partition.id
+      - ref: messaging.destination.name
+        requirement_level:
+          conditionally_required: if and only if `messaging.destination.name` is known to have low cardinality. Otherwise, `messaging.destination.template` MAY be populated.
+      - ref: messaging.destination.template
+        requirement_level:
+          conditionally_required: if available.
+
+  # durations
+  - id: metric.messaging.client.request.duration
+    type: metric
+    metric_name: messaging.client.request.duration
+    brief: "Duration of messaging operation initiated by a producer or consumer."
+    note: >
+      This metric SHOULD NOT be used to report processing duration - processing duration is reported in `messaging.process.duration` metric.
+    stability: experimental
+    instrument: histogram
+    unit: "s"
+    extends: metric.messaging.attributes
+    attributes:
+      - ref: messaging.operation.type
+        requirement_level:
+          conditionally_required: If applicable.
+      - ref: messaging.operation.name
+        examples: ["send", "receive", "ack"]
+
+  - id: metric.messaging.process.duration
+    type: metric
+    metric_name: messaging.process.duration
+    brief: "Duration of processing operation."
+    note: >
+      This metric MUST be reported for operations with `messaging.operation.type` that matches `process`.
+    stability: experimental
+    instrument: histogram
+    unit: "s"
+    extends: metric.messaging.attributes
+    attributes:
+      - ref: messaging.operation.name
+        examples: ["process", "consume", "handle"]
+
+  # counters
+  - id: metric.messaging.client.published.messages
+    type: metric
+    metric_name: messaging.client.published.messages
+    brief: "Number of messages producer attempted to publish to the broker."
+    note: >
+      This metric MUST NOT count messages that were created, but aren't being published yet.
+    stability: experimental
+    instrument: counter
+    unit: "{message}"
+    extends: metric.messaging.attributes
+    attributes:
+      - ref: messaging.operation.name
+        examples: ["send", "schedule", "enqueue"]
+
+  - id: metric.messaging.client.consumed.messages
+    type: metric
+    metric_name: messaging.client.consumed.messages
+    brief: "Number of messages that were delivered to the application."
+    note: >
+      Records the number of messages pulled from the broker or number of messages dispatched to the application in push-based scenarios.
+
+      The metric MUST be reported once per message delivery. For example, if receiving and processing operations are both instrumented for a single message delivery, the number of received messages
+      is reported when then message is received and not reported for the processing operation.
```

```suggestion
    brief: "Number of messages that were forwarded to the application."
    note: >
      Records the number of messages pulled from the broker or number of messages dispatched to the application in push-based scenarios.

      The metric MUST be reported once for each forwarded message. For example, if receiving and processing operations are both instrumented for a single message delivery, the number of received messages
      is reported when then message is received and not reported for the processing operation.
```

Maybe avoiding the term "delivery", as it connotates push-based scenarios?

---

I use it it to emphasize that it's not per message, but per message delivery. A message can be received/delivered multiple times.

'Delivery' is used in general messaging scenarios without pull/push assumption - e.g. 
- JMS - https://javaee.github.io/javaee-spec/javadocs/javax/jms/Message.html#getJMSRedelivered--
- ActiveMQ - https://activemq.apache.org/components/artemis/documentation/latest/undelivered-messages
- RabbitMQ - https://www.rabbitmq.com/docs/consumers#message-properties

---

Makes sense. I was wondering about different terms we're using ("consume", "receive", "deliver", "forward") and being clear about the usage. It makes sense to use "deliver" for the whole process of getting a message from the broker to the application code, regardless of pull/push scenarios.